### PR TITLE
[buienradar] Fixes for lat/lon, retry logic and timestamp channel

### LINKED ISSUE #5718: [buienradar] binding only reports 23 of the 24 available forecast values
Although the binding correctly processes the 24 forecast values reported by the Buienradar web API, only 23 are propagated to OH.

Version: openHAB 2.5.0 Build #1601

Also the documentation states that there are 25 values available (current + 24 forecast), which is not possible.

I suppose there is an off-by-5-minutes bug in the binding.

See https://community.openhab.org/t/buienradar-binding/74954/12

### LINKED ISSUE #5719: [buienradar] if prediction retrieval fails, exception occurs and no retry happens
I’m running the binding on 2.5.0-SNAPSHOT (build #1601).

The buienradar binding seems to run occasionally in concurrent execution issues:
```
2019-06-12 17:30:27.105 [WARN ] [uienradar.internal.BuienradarHandler] - Cannot retrieve predictions
java.io.IOException: java.util.concurrent.ExecutionException: java.io.EOFException: HttpConnectionOverHTTP@1264519::DecryptedEndPoint@6415ee{gpsgadget.buienradar.nl/173.223.106.128:443<->/192.168.0.10:53068,OPEN,fill=-,flush=-,to=300154/0}
        at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrlAndGetReponse(HttpUtil.java:259) ~[?:?]
        at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrl(HttpUtil.java:156) ~[?:?]
        at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrl(HttpUtil.java:131) ~[?:?]
        at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrl(HttpUtil.java:109) ~[?:?]
        at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrl(HttpUtil.java:90) ~[?:?]
        at org.openhab.binding.buienradar.internal.buienradarapi.BuienradarPredictionAPI.getPredictions(BuienradarPredictionAPI.java:133) ~[?:?]
        at org.openhab.binding.buienradar.internal.BuienradarHandler.refresh(BuienradarHandler.java:108) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.util.concurrent.ExecutionException: java.io.EOFException: HttpConnectionOverHTTP@1264519::DecryptedEndPoint@6415ee{gpsgadget.buienradar.nl/173.223.106.128:443<->/192.168.0.10:53068,OPEN,fill=-,flush=-,to=300154/0}
        at org.eclipse.jetty.client.util.FutureResponseListener.getResult(FutureResponseListener.java:118) ~[?:?]
        at org.eclipse.jetty.client.util.FutureResponseListener.get(FutureResponseListener.java:101) ~[?:?]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:684) ~[?:?]
        at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrlAndGetReponse(HttpUtil.java:250) ~[?:?]
        ... 13 more
Caused by: java.io.EOFException: HttpConnectionOverHTTP@1264519::DecryptedEndPoint@6415ee{gpsgadget.buienradar.nl/173.223.106.128:443<->/192.168.0.10:53068,OPEN,fill=-,flush=-,to=300154/0}
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.earlyEOF(HttpReceiverOverHTTP.java:338) ~[?:?]
        at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1551) ~[?:?]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.shutdown(HttpReceiverOverHTTP.java:209) ~[?:?]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:147) ~[?:?]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:73) ~[?:?]
        at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:133) ~[?:?]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:155) ~[?:?]
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305) ~[?:?]
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) ~[?:?]
        at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:411) ~[?:?]
        at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:305) ~[?:?]
        at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:159) ~[?:?]
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) ~[?:?]
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118) ~[?:?]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765) ~[?:?]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683) ~[?:?]
        ... 1 more
```
Ideally, if such exception happens, a new attempt is scheduled (e.g., 1 minute later).

### LINKED ISSUE #5723: [buienradar] Expose timestamp of retrieval (or first prediction) as channel
The buienradar binding exposes all 24 rain forecast values to OH but it doesn't yet expose the time of retrieval or the time of the first prediction (2nd parameter on the 1st line returned by the buienradar web API). This information can be relevant for tracking when a forecast update was performed, e.g., in a sitemap or in a rule.

### LINKED ISSUE #5735: [buienradar] Unhandled java.util.concurrent.TimeoutException (Buienradar API frequently takes longer than 3000ms to reply)
The buienradar binding expects the Buienradar API to return an HTTP request in 3000 milliseconds (see **BuienradarPredictionAPI.java**). However, the Buienradar API often takes longer than the timeout value before it is able to handle the request. In such cases, an unhandled exception is thrown:

> 2019-06-16 00:47:38.618 [WARN ] [uienradar.internal.BuienradarHandler] - Cannot retrieve predictions
> **java.io.IOException: java.util.concurrent.TimeoutException**: Total timeout 3000 ms elapsed
> 	at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrlAndGetReponse(HttpUtil.java:259) ~[?:?]
> 	at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrl(HttpUtil.java:156) ~[?:?]
> 	at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrl(HttpUtil.java:131) ~[?:?]
> 	at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrl(HttpUtil.java:109) ~[?:?]
> 	at org.eclipse.smarthome.io.net.http.HttpUtil.executeUrl(HttpUtil.java:90) ~[?:?]
> 	at org.openhab.binding.buienradar.internal.buienradarapi.BuienradarPredictionAPI.getPredictions(**BuienradarPredictionAPI.java:133**) ~[?:?]
> 	at org.openhab.binding.buienradar.internal.BuienradarHandler.refresh(BuienradarHandler.java:108) ~[?:?]
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
> 	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [?:?]
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [?:?]
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [?:?]
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
> 	at java.lang.Thread.run(Thread.java:748) [?:?]
> Caused by: java.util.concurrent.TimeoutException: Total timeout 3000 ms elapsed
> 	at org.eclipse.jetty.client.TimeoutCompleteListener.onTimeoutExpired(TimeoutCompleteListener.java:52) ~[?:?]
> 	at org.eclipse.jetty.io.CyclicTimeout$Wakeup.run(CyclicTimeout.java:271) ~[?:?]
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:?]
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:?]
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:?]
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:?]
> 	... 3 more

Expected behavior: either one of:

1. Quick hack: Increase the timeout (e.g., set it to 5000 ms)
2. More elaborative solution: If the `java.util.concurrent.TimeoutException` is thrown (response not received within the predefined time), catch it by backing off for a couple seconds and retrying it (up to say max 3 times before bailing out). Optionally the back-off time interval could double after each retry.

### LINKED ISSUE #5745: [buienradar] Incorrect lat/lon format results in unnecessary redirects when fetching forecast data
The buienradar widget makes use of the following web service URI: `gpsgadget.buienradar.nl/data/raintext?lat=XX.XXXXX&lon=YY.YYYYY`

When manually fetching the forecast info, I noticed that the web service redirects to (lat, lon) coodinates with 2-digit precision, i.e.: `gpsgadget.buienradar.nl/data/raintext?lat=XX.XX&lon=YY.YY`

This may explain why the buienradar binding frequently misses out on fetching forecast info.

Discussion at OH Community: https://community.openhab.org/t/buienradar-binding-store-lat-lon-with-2-digit-precision/76632
 
## Expected Behavior
Avoid the redirect by constructing the URI with 2-digit precision for lat/lon.

## Current Behavior
Currently lat/lon are provided with 8-digit precision, see **BuienradarPredictionAPI.java**, line 131:
```
    @Override
    public List<Prediction> getPredictions(PointType location) throws IOException {
        final String address = String.format(Locale.ENGLISH, BASE_ADDRESS + "?lat=%.8f&lon=%.8f",
                location.getLatitude().doubleValue(), location.getLongitude().doubleValue());
```

## Possible Solution

Edit the `address` definition as follows:
```
        final String address = String.format(Locale.ENGLISH, BASE_ADDRESS + "?lat=%.2f&lon=%.2f",
```
## Workaround

Provide the location info with 2-digit precision in the binding config (may still yield redirects due to floating point conversion / accuracy).

## Steps to Reproduce (for Bugs)
Perform the HTTPS request by hand, e.g. with `wget -O - https://gpsgadget.buienradar.nl/data/raintext...` and confirm the redirect.

## Context
The buienradar binding very frequently times out. It expect a response within 3000 ms. If an avoidable redirect can be removed, (1) the binding will less frequently time out and (2) the buienradar web API will be less hammered.

## Your Environment
OH2.5.0 snapshot build S1601 on RPi3B+ with OpenHABian.

### PR DESCRIPTION
**Pull request by gedejong**
Fixes #5745 [buienradar] Incorrect lat/lon format results in unnecessary redirects when fetching forecast data
Fixes #5735 [buienradar] Unhandled java.util.concurrent.TimeoutException (Buienradar API frequently takes longer than 3000ms to reply)
Fixes #5723 [buienradar] Expose timestamp of retrieval (or first prediction) as channel
Fixes #5719 [buienradar] if prediction retrieval fails, exception occurs and no retry happens
Fixes #5718 [buienradar] binding only reports 23 of the 24 available forecast values

Due to relatedness of above bugs, this is placed in one commit.

Also updated the documentation to include an example Grafana dashboard.

### CODE REVIEW

File: bundles/org.openhab.binding.buienradar/README.md [hunk 1]
```diff
@@ -50,3 +55,20 @@ Number RAIN_115MIN "Rain 115 min." (Rain) {channel="buienradar:rain_forecast:hom
 Number RAIN_120MIN "Rain 120 min." (Rain) {channel="buienradar:rain_forecast:home:forecast_120" }
 ```
 
+## Example data visualisation
+
+In dit voorbeeld gebruiken we de 'Discrete' plugin van Grafana om een visualisatie te maken van de voorspellingen. Er zijn mappings gemaakt naar neerslagintensiteit (licht, middelmatig, zwaar) en die categorieen krijgen de desbetreffende kleuren.
```
Comment (J-N-K) [inline bundles/org.openhab.binding.buienradar/README.md]: Please use english here.

File: bundles/org.openhab.binding.buienradar/README.md [hunk 2]
```diff
@@ -50,3 +55,20 @@ Number RAIN_115MIN "Rain 115 min." (Rain) {channel="buienradar:rain_forecast:hom
...
+
+![Z-Way Binding](doc/img/grafana-dashboard.png)
+
+De mappings zijn als volgt:
```
Comment (J-N-K) [inline bundles/org.openhab.binding.buienradar/README.md]: see above

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java [hunk 3]
```diff
@@ -54,15 +57,15 @@
 
     private final Logger logger = LoggerFactory.getLogger(BuienradarHandler.class);
 
+    private static final Unit<Speed> MILLIMETRE_PER_HOUR = Units.METRE.divide(1000).divide(Units.HOUR)
```
Comment (J-N-K) [inline bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java]: Why don't we add that to the smarthome units?

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java [hunk 4]
```diff
@@ -97,25 +100,51 @@ public void initialize() {
             updateStatus(ThingStatus.UNKNOWN);
         }
         if (listenableFuture == null || listenableFuture.isCancelled()) {
-            listenableFuture = scheduler.scheduleWithFixedDelay(this::refresh, 0L, config.refreshIntervalMinutes,
-                    MINUTES);
+            listenableFuture = scheduler.scheduleWithFixedDelay(() -> refresh(config), 0L,
```
Comment (J-N-K) [inline bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java]: Wouldn't it be easier to store the configuration in the handler and use it when you need it?

File: bundles/org.openhab.binding.buienradar/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 5]
```diff
@@ -50,9 +51,28 @@
...
 				<default>5</default>
 			</parameter>
+			
+			<parameter name="retries" type="integer" required="true" min="1" >
+                <label>Retries</label>
+                <description>Number of retries to try to retrieve buienradar results</description>
+                <default>4</default>
+            </parameter>
+			
+			<parameter name="exponentialBackoffRetryBaseInSeconds" type="integer" required="true" min="1" unit="s">
```
Comment (J-N-K) [inline bundles/org.openhab.binding.buienradar/src/main/resources/ESH-INF/thing/thing-types.xml]: Please check the formatting in this file

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java (hunk 3 above)
Comment (gedejong) [reply bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java]: It is already added. However, I cannot test this binding on my RPi because it is only in 2.5.0-SNAPSHOT.

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java (hunk 4 above)
Comment (gedejong) [reply bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java]: I am trying to keep mutable state in the handler to a minimum. Now it is easy to prove that `initialize()` will always re-read the configuration.

File: bundles/org.openhab.binding.buienradar/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 5 above)
Comment (gedejong) [reply bundles/org.openhab.binding.buienradar/src/main/resources/ESH-INF/thing/thing-types.xml]: Ah, I thought teh xml files would also auto-format on save...

File: bundles/org.openhab.binding.buienradar/README.md (hunk 2 above)
Comment (gedejong) [reply bundles/org.openhab.binding.buienradar/README.md]: Would you believe I copy-pasted the wrong side of the [deepl translator](deepl.com)? Fixed anyways.

**Comment by gedejong** [comment]
Build is failing due to another binding.

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java (hunk 4 above)
Comment (gedejong) [reply bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java]: I guess I'm used to an FP, immutability kind of thinking. 'config' as a field in your proposal would need to be a mutable and nullable field, making it harder to reason whether config is accessed from the right methods (from `initialize` instead of many of the other methods). This places `config` in the lambda closure, as if I had created a new `Runnable` with one field initialized in the constructor called `config`.  It prevents needless `null`-checks, possible race-conditions and reduced efferent coupling with other state.

So, yes, I considered making `config` a field, but decided against it, since it is only one field, it is the defining part of the closure (no other fields needed) and (imho) made the code more readable.

But if you disagree... Well, the code needs to be merged and it would make a large number of people happy, so I'm more than happy to have you make the suggested change.

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java (hunk 3 above)
Comment (gedejong) [reply bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java]: That would make testing a lot harder for me (and other contributors). I mean, essentially we are programming the bindings against the 2.4.0 API, right? There is no forced upgrade path, the API is backward compatible. Otherwise we'd have an ordering problem in the commits as well.

For now (I don't want to argue endlessly), I'll just fix this opaque, almost undocumented call to the `Units` API with `MILLIMETRE_PER_HOUR` from the arbitrary mishmash of UoM in `SmartHomeUnits`

Btw, where is the documentation of the uom library? The [github](https://github.com/unitsofmeasurement/uom-se) page has a minimal `README.md`, JSR 363 I have no idea where to begin (don't click on the first link in the list below, `http://www.unitsofmeasurement.org/` will redirect to a spam domain), even Martin Fowlers site on quantities is no longer online (https://martinfowler.com/eaaDev/quantity.html) and Googling I cannot find a good list of examples (except in Kotlin...)

Don't get me wrong, I am *all* *for* type-safety when dealing with quantities. It's just that I am spoiled with Scala's Squants.

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java (hunk 4 above)
Comment (gedejong) [reply bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java]: Correction, I am more than happy to make that change.

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java [hunk 6]
```diff
@@ -133,9 +171,14 @@ private void refresh() {
     @SuppressWarnings("null")
     @Override
     public void dispose() {
-        if (listenableFuture != null && !listenableFuture.isCancelled()) {
-            listenableFuture.cancel(true);
-            listenableFuture = null;
+        try {
+            listenableFutureLock.lock();
```
Comment (J-N-K) [inline bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java]: I don‘t think this hurts, but: why did you add the lock here?

**Comment by J-N-K** [comment]
Waiting for Travis.

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java [hunk 7]
```diff
@@ -96,29 +103,60 @@ public void initialize() {
...
-                    final String label = String.format(Locale.ENGLISH, "forecast_%d", minuteClass);
+
+                final long minutesFromNow = prediction.getActualDateTime().until(prediction.getDateTimeOfPrediction(),
+                        ChronoUnit.MINUTES);
+                logger.debug("Forecast for {} at {} made at {} is {}", minutesFromNow,
+                        prediction.getDateTimeOfPrediction(), prediction.getActualDateTime(), intensity);
+                if (minutesFromNow >= 0 && minutesFromNow <= 115) {
+                    final String label = String.format(Locale.ENGLISH, "forecast_%d", minutesFromNow);
 
                     /** @TODO: edejong 2019-04-03 Change to SmartHomeUnits.MILLIMETRE_PER_HOUR for OH 2.5 */
```
Comment (martinvw) [inline bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java]: Actually you already resolved this PR, but I won't block it for this

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java [hunk 8]
```diff
@@ -120,37 +123,53 @@ public final BigDecimal getIntensity() {
...
 
             @Override
-            public final ZonedDateTime getDateTime() {
+            public ZonedDateTime getDateTimeOfPrediction() {
                 return dateTime;
             }
+
+            @Override
+            public ZonedDateTime getActualDateTime() {
+                // TODO Auto-generated method stub
```
Comment (martinvw) [inline bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java]: But I dont like this TODO / comment :-)

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java (hunk 6 above)
Comment (gedejong) [reply bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/BuienradarHandler.java]: Because I cannot sync on a nullable value (listenableFuture) and this makes the class thread-safe (initialize / dispose both read/write listenableFuture).

File: bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java (hunk 8 above)
Comment (gedejong) [reply bundles/org.openhab.binding.buienradar/src/main/java/org/openhab/binding/buienradar/internal/buienradarapi/BuienradarPredictionAPI.java]: grmbl... This really brings back memories to all the crappy auto-generated comments everywhere. I removed it.
